Enables backend preprocessing to take place outside of the backend interface

raziel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51757 Enables backend preprocessing to take place outside of the backend interface**

Enables backend preprocessing to take place outside of the backend interface.

What's new:
* A new definition for backend preprocessing is defined (i.e. BackendPreprocessFunction).
* Registration of the backend's PyTorchBackendInterface interface implementation is augmented to take the BackendPreprocessFunction.
* A new registry is created to handle the BackendPreprocessFunction functions, using the backend's name as key.
* When a BackendPreprocessFunction is used, the  PyTorchBackendInterface's "preprocess" method is not added to the LoweredModule. Instead, the BackendPreprocessFunction is called and its output used to set the LoweredModule's __processed_module.


Why?:
These changes are needed to avoid forcing "preprocessing" part of the LoweredModule, and in the future of PyTorchBackendInterface.
This is important for Mobile use cases where "preprocess" can take the bulk of the compilation process, and thus contain code dependencies that we do not want to bring (or cannot bring) to the Mobile binary.

What didn't change:
* Everything is backwards compatible:
** The existing "preprocess" method in PyTorchBackendInterface.
** When backend registration is done without the BackendPreprocessFunction, as before, things work the same way: "preprocess" is added to LoweredModule, and invoked through the module's instance of the backend interface.

Longer term, the plan is to refactor existing users to move to the new backend registration.

Differential Revision: [D26261042](https://our.internmc.facebook.com/intern/diff/D26261042/)

Commit:
Enables backend preprocessing to take place outside of the backend interface

Enables backend preprocessing to take place outside of the backend interface.

What's new:
* A new definition for backend preprocessing is defined (i.e. BackendPreprocessFunction).
* Registration of the backend's PyTorchBackendInterface interface implementation is augmented to take the BackendPreprocessFunction.
* A new registry is created to handle the BackendPreprocessFunction functions, using the backend's name as key.
* When a BackendPreprocessFunction is used, the  PyTorchBackendInterface's "preprocess" method is not added to the LoweredModule. Instead, the BackendPreprocessFunction is called and its output used to set the LoweredModule's __processed_module.


Why?:
These changes are needed to avoid forcing "preprocessing" part of the LoweredModule, and in the future of PyTorchBackendInterface.
This is important for Mobile use cases where "preprocess" can take the bulk of the compilation process, and thus contain code dependencies that we do not want to bring (or cannot bring) to the Mobile binary.

What didn't change:
* Everything is backwards compatible:
** The existing "preprocess" method in PyTorchBackendInterface.
** When backend registration is done without the BackendPreprocessFunction, as before, things work the same way: "preprocess" is added to LoweredModule, and invoked through the module's instance of the backend interface.

Longer term, the plan is to refactor existing users to move to the new backend registration.

Differential Revision: [D26261042](https://our.internmc.facebook.com/intern/diff/D26261042/)

[ghstack-poisoned]

Commit:
Update on "Enables backend preprocessing to take place outside of the backend interface"


Enables backend preprocessing to take place outside of the backend interface.

What's new:
* A new definition for backend preprocessing is defined (i.e. BackendPreprocessFunction).
* Registration of the backend's PyTorchBackendInterface interface implementation is augmented to take the BackendPreprocessFunction.
* A new registry is created to handle the BackendPreprocessFunction functions, using the backend's name as key.
* When a BackendPreprocessFunction is used, the  PyTorchBackendInterface's "preprocess" method is not added to the LoweredModule. Instead, the BackendPreprocessFunction is called and its output used to set the LoweredModule's __processed_module.


Why?:
These changes are needed to avoid forcing "preprocessing" part of the LoweredModule, and in the future of PyTorchBackendInterface.
This is important for Mobile use cases where "preprocess" can take the bulk of the compilation process, and thus contain code dependencies that we do not want to bring (or cannot bring) to the Mobile binary.

What didn't change:
* Everything is backwards compatible:
** The existing "preprocess" method in PyTorchBackendInterface.
** When backend registration is done without the BackendPreprocessFunction, as before, things work the same way: "preprocess" is added to LoweredModule, and invoked through the module's instance of the backend interface.

Longer term, the plan is to refactor existing users to move to the new backend registration.

Differential Revision: [D26261042](https://our.internmc.facebook.com/intern/diff/D26261042/)

[ghstack-poisoned]

Commit:
Update on "Enables backend preprocessing to take place outside of the backend interface"


Enables backend preprocessing to take place outside of the backend interface.

What's new:
* A new definition for backend preprocessing is defined (i.e. BackendPreprocessFunction).
* Registration of the backend's PyTorchBackendInterface interface implementation is augmented to take the BackendPreprocessFunction.
* A new registry is created to handle the BackendPreprocessFunction functions, using the backend's name as key.
* When a BackendPreprocessFunction is used, the  PyTorchBackendInterface's "preprocess" method is not added to the LoweredModule. Instead, the BackendPreprocessFunction is called and its output used to set the LoweredModule's __processed_module.


Why?:
These changes are needed to avoid forcing "preprocessing" part of the LoweredModule, and in the future of PyTorchBackendInterface.
This is important for Mobile use cases where "preprocess" can take the bulk of the compilation process, and thus contain code dependencies that we do not want to bring (or cannot bring) to the Mobile binary.

What didn't change:
* Everything is backwards compatible:
** The existing "preprocess" method in PyTorchBackendInterface.
** When backend registration is done without the BackendPreprocessFunction, as before, things work the same way: "preprocess" is added to LoweredModule, and invoked through the module's instance of the backend interface.

Longer term, the plan is to refactor existing users to move to the new backend registration.

Differential Revision: [D26261042](https://our.internmc.facebook.com/intern/diff/D26261042/)

[ghstack-poisoned]

Commit:
Update on "Enables backend preprocessing to take place outside of the backend interface"


Enables backend preprocessing to take place outside of the backend interface.

What's new:
* A new definition for backend preprocessing is defined (i.e. BackendPreprocessFunction).
* Registration of the backend's PyTorchBackendInterface interface implementation is augmented to take the BackendPreprocessFunction.
* A new registry is created to handle the BackendPreprocessFunction functions, using the backend's name as key.
* When a BackendPreprocessFunction is used, the  PyTorchBackendInterface's "preprocess" method is not added to the LoweredModule. Instead, the BackendPreprocessFunction is called and its output used to set the LoweredModule's __processed_module.


Why?:
These changes are needed to avoid forcing "preprocessing" part of the LoweredModule, and in the future of PyTorchBackendInterface.
This is important for Mobile use cases where "preprocess" can take the bulk of the compilation process, and thus contain code dependencies that we do not want to bring (or cannot bring) to the Mobile binary.

What didn't change:
* Everything is backwards compatible:
** The existing "preprocess" method in PyTorchBackendInterface.
** When backend registration is done without the BackendPreprocessFunction, as before, things work the same way: "preprocess" is added to LoweredModule, and invoked through the module's instance of the backend interface.

Longer term, the plan is to refactor existing users to move to the new backend registration.

Differential Revision: [D26261042](https://our.internmc.facebook.com/intern/diff/D26261042/)

[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -329,13 +349,6 @@ void initJitBackendBindings(PyObject* module) {
 
-    // Run preprocess so that __processed_module is set correctly before
-    // compilation.
-    loweredModule.run_method(
```
I'm wondering how is it backward compatible without running __preprocess, when there's no BackendPreprocessFunction? In such a case, would this piece of code still be needed?

iseeyuan commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -153,11 +153,18 @@ void initJitBackendBindings(PyObject* module) {
         AnyType::get(),
-        cloned_module._ivalue(),
+        detail::hasBackendPreprocessFunction(backend_name)
+            ? detail::getBackendPreprocessFunction(backend_name)(
```
Aha, the compilation can happen here.

raziel replied:
That's right.
Here we set the value by calling the BackendPreprocessFunction (if available).
Otherwise we do as before and first set the cloned original module, and then below we add the __preprocess function and call it (as before).

raziel replied:
Discussed how this works offline.
Resolving the comment.

SplitInfinity commented on torch/csrc/jit/backends/backend.h:
```diff
@@ -14,10 +15,11 @@ template <class TBackendInterface>
   std::string backend_name_;
 
  public:
+  C10_DEPRECATED
```
Oh cool, how does this work?

SplitInfinity replied:
Maybe we should mark preprocess on `PyTorchBackendInterface` as deprecated also?

raziel replied:
It's our in-house macro for the equivalent C++14 [[deprecated]] that handles portability across platforms.
Added it to PyTorchBackendInterface as suggested.

SplitInfinity commented on torch/csrc/jit/backends/backend_detail.cpp:
```diff
@@ -48,6 +50,38 @@ c10::FunctionSchema getExecuteSchema() {
+    const BackendPreprocessFunction& preprocess) {
+  TORCH_CHECK(
+      !detail::hasBackendPreprocessFunction(name),
+      "BackendPreprocessFunction for backend ",
```
```suggestion
      "Preprocessing function for backend ",
```

SplitInfinity commented on test/cpp/jit/test_backend.cpp:
```diff
@@ -67,8 +67,15 @@ class TestBackend : public PyTorchBackendInterface {
   }
 };
 
+c10::IValue backendPreprocessFunction(
```
```suggestion
c10::IValue preprocess(
```

SplitInfinity commented on torch/csrc/jit/backends/backend_detail.h:
```diff
@@ -46,6 +50,20 @@ std::function<void(Stack&)> getExecuteFunc() {
   };
 }
+
+typedef std::function<
+    c10::IValue(const Module&, const c10::Dict<IValue, IValue>&)>
+    BackendPreprocessFunction;
```
```suggestion
using BackendPreprocessFunction = std::function<
    c10::IValue(const Module&, const c10::Dict<IValue, IValue>&)>;
```

SplitInfinity commented on torch/csrc/jit/backends/backend_detail.h:
```diff
@@ -46,6 +50,20 @@ std::function<void(Stack&)> getExecuteFunc() {
+    c10::IValue(const Module&, const c10::Dict<IValue, IValue>&)>
+    BackendPreprocessFunction;
+
+TORCH_API void registerBackendPreprocessFunction(
+    const std::string& name,
+    const BackendPreprocessFunction& preprocess);
+
+TORCH_API bool hasBackendPreprocessFunction(const std::string& name);
+
+TORCH_API BackendPreprocessFunction
```
Do these need to be declared with `TORCH_API`?

raziel replied:
That's right.
Just registerBackendPreprocessFunction
I removed it from the others.

raziel replied:
Afterall, the 3 functions need to be exported with TORCH_API since:
* backend.h uses registerBackendPreprocessFunction() and backend_init.cpp uses hasBackendPreprocessFunction() and getBackendPreprocessFunction(), AND
* backend_detail.h/.cpp is in a separate .so (libtorch) from backend.h/backend_init.h/.cpp (libtorch_python)

SplitInfinity replied:
Thanks for following up.

For my own knowledge, what does the macro do, exactly?

SplitInfinity commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -190,9 +197,9 @@ void initJitBackendBindings(PyObject* module) {
     // This is a helper function for creating a new instance of the
     // backend class.
     static const auto create_backend_ct = CodeTemplate(R"(
-            def __create_backend(self):
-                self.__backend = $name()
-            )");
+              def __create_backend(self):
+                  self.__backend = $name()
+              )");
```
Intentional formatting changes?

raziel replied:
yes, or so the VSCode said so.

SplitInfinity commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -202,44 +209,57 @@ void initJitBackendBindings(PyObject* module) {
     // the LoweredModule.
     loweredModule.define(
         R"(
-            def __getstate__(self):
-                return self.__method_compile_spec, self.__processed_module
-            )",
+              def __getstate__(self):
+                  return self.__method_compile_spec, self.__processed_module
+              )",
```
Intentional formatting changes?

raziel replied:
fixed.

SplitInfinity commented on torch/csrc/jit/backends/backend_detail.cpp:
```diff
@@ -48,6 +50,38 @@ c10::FunctionSchema getExecuteSchema() {
       /*returns=*/{output});
 }
+
+std::unordered_map<std::string, BackendPreprocessFunction>&
+backendPreprocessFunctions() {
+  static std::unordered_map<std::string, BackendPreprocessFunction>
+      preprocess_functions;
+  return preprocess_functions;
+}
```
```suggestion
namespace {
std::unordered_map<std::string, BackendPreprocessFunction>&
backendPreprocessFunctions() {
  static std::unordered_map<std::string, BackendPreprocessFunction>
      preprocess_functions;
  return preprocess_functions;
}
}
```

SplitInfinity replied:
To make sure this is visible only in this translation unit and can only be accessed through the methods below.

SplitInfinity reviewed (commented):
Nice! I don't see any major issues, but I'd like a response to some of my suggestions and questions before I accept.

Commit:
Update on "Enables backend preprocessing to take place outside of the backend interface"


Enables backend preprocessing to take place outside of the backend interface.

What's new:
* A new definition for backend preprocessing is defined (i.e. BackendPreprocessFunction).
* Registration of the backend's PyTorchBackendInterface interface implementation is augmented to take the BackendPreprocessFunction.
* A new registry is created to handle the BackendPreprocessFunction functions, using the backend's name as key.
* When a BackendPreprocessFunction is used, the  PyTorchBackendInterface's "preprocess" method is not added to the LoweredModule. Instead, the BackendPreprocessFunction is called and its output used to set the LoweredModule's __processed_module.


Why?:
These changes are needed to avoid forcing "preprocessing" part of the LoweredModule, and in the future of PyTorchBackendInterface.
This is important for Mobile use cases where "preprocess" can take the bulk of the compilation process, and thus contain code dependencies that we do not want to bring (or cannot bring) to the Mobile binary.

What didn't change:
* Everything is backwards compatible:
** The existing "preprocess" method in PyTorchBackendInterface.
** When backend registration is done without the BackendPreprocessFunction, as before, things work the same way: "preprocess" is added to LoweredModule, and invoked through the module's instance of the backend interface.

Longer term, the plan is to refactor existing users to move to the new backend registration.

Differential Revision: [D26261042](https://our.internmc.facebook.com/intern/diff/D26261042/)

[ghstack-poisoned]

Commit:
Update on "Enables backend preprocessing to take place outside of the backend interface"


Enables backend preprocessing to take place outside of the backend interface.

What's new:
* A new definition for backend preprocessing is defined (i.e. BackendPreprocessFunction).
* Registration of the backend's PyTorchBackendInterface interface implementation is augmented to take the BackendPreprocessFunction.
* A new registry is created to handle the BackendPreprocessFunction functions, using the backend's name as key.
* When a BackendPreprocessFunction is used, the  PyTorchBackendInterface's "preprocess" method is not added to the LoweredModule. Instead, the BackendPreprocessFunction is called and its output used to set the LoweredModule's __processed_module.


Why?:
These changes are needed to avoid forcing "preprocessing" part of the LoweredModule, and in the future of PyTorchBackendInterface.
This is important for Mobile use cases where "preprocess" can take the bulk of the compilation process, and thus contain code dependencies that we do not want to bring (or cannot bring) to the Mobile binary.

What didn't change:
* Everything is backwards compatible:
** The existing "preprocess" method in PyTorchBackendInterface.
** When backend registration is done without the BackendPreprocessFunction, as before, things work the same way: "preprocess" is added to LoweredModule, and invoked through the module's instance of the backend interface.

Longer term, the plan is to refactor existing users to move to the new backend registration.

Differential Revision: [D26261042](https://our.internmc.facebook.com/intern/diff/D26261042/)

[ghstack-poisoned]

Commit:
Update on "Enables backend preprocessing to take place outside of the backend interface"


Enables backend preprocessing to take place outside of the backend interface.

What's new:
* A new definition for backend preprocessing is defined (i.e. BackendPreprocessFunction).
* Registration of the backend's PyTorchBackendInterface interface implementation is augmented to take the BackendPreprocessFunction.
* A new registry is created to handle the BackendPreprocessFunction functions, using the backend's name as key.
* When a BackendPreprocessFunction is used, the  PyTorchBackendInterface's "preprocess" method is not added to the LoweredModule. Instead, the BackendPreprocessFunction is called and its output used to set the LoweredModule's __processed_module.


Why?:
These changes are needed to avoid forcing "preprocessing" part of the LoweredModule, and in the future of PyTorchBackendInterface.
This is important for Mobile use cases where "preprocess" can take the bulk of the compilation process, and thus contain code dependencies that we do not want to bring (or cannot bring) to the Mobile binary.

What didn't change:
* Everything is backwards compatible:
** The existing "preprocess" method in PyTorchBackendInterface.
** When backend registration is done without the BackendPreprocessFunction, as before, things work the same way: "preprocess" is added to LoweredModule, and invoked through the module's instance of the backend interface.

Longer term, the plan is to refactor existing users to move to the new backend registration.

Differential Revision: [D26261042](https://our.internmc.facebook.com/intern/diff/D26261042/)

[ghstack-poisoned]

SplitInfinity reviewed (approved):
Nice! Let's keep an eye on test signals from internal use cases.

Commit:
Update on "Enables backend preprocessing to take place outside of the backend interface"


Enables backend preprocessing to take place outside of the backend interface.

What's new:
* A new definition for backend preprocessing is defined (i.e. BackendPreprocessFunction).
* Registration of the backend's PyTorchBackendInterface interface implementation is augmented to take the BackendPreprocessFunction.
* A new registry is created to handle the BackendPreprocessFunction functions, using the backend's name as key.
* When a BackendPreprocessFunction is used, the  PyTorchBackendInterface's "preprocess" method is not added to the LoweredModule. Instead, the BackendPreprocessFunction is called and its output used to set the LoweredModule's __processed_module.


Why?:
These changes are needed to avoid forcing "preprocessing" part of the LoweredModule, and in the future of PyTorchBackendInterface.
This is important for Mobile use cases where "preprocess" can take the bulk of the compilation process, and thus contain code dependencies that we do not want to bring (or cannot bring) to the Mobile binary.

What didn't change:
* Everything is backwards compatible:
** The existing "preprocess" method in PyTorchBackendInterface.
** When backend registration is done without the BackendPreprocessFunction, as before, things work the same way: "preprocess" is added to LoweredModule, and invoked through the module's instance of the backend interface.

Longer term, the plan is to refactor existing users to move to the new backend registration.

Differential Revision: [D26261042](https://our.internmc.facebook.com/intern/diff/D26261042/)

[ghstack-poisoned]

Commit:
Update on "Enables backend preprocessing to take place outside of the backend interface"


Enables backend preprocessing to take place outside of the backend interface.

What's new:
* A new definition for backend preprocessing is defined (i.e. BackendPreprocessFunction).
* Registration of the backend's PyTorchBackendInterface interface implementation is augmented to take the BackendPreprocessFunction.
* A new registry is created to handle the BackendPreprocessFunction functions, using the backend's name as key.
* When a BackendPreprocessFunction is used, the  PyTorchBackendInterface's "preprocess" method is not added to the LoweredModule. Instead, the BackendPreprocessFunction is called and its output used to set the LoweredModule's __processed_module.


Why?:
These changes are needed to avoid forcing "preprocessing" part of the LoweredModule, and in the future of PyTorchBackendInterface.
This is important for Mobile use cases where "preprocess" can take the bulk of the compilation process, and thus contain code dependencies that we do not want to bring (or cannot bring) to the Mobile binary.

What didn't change:
* Everything is backwards compatible:
** The existing "preprocess" method in PyTorchBackendInterface.
** When backend registration is done without the BackendPreprocessFunction, as before, things work the same way: "preprocess" is added to LoweredModule, and invoked through the module's instance of the backend interface.

Longer term, the plan is to refactor existing users to move to the new backend registration.

Differential Revision: [D26261042](https://our.internmc.facebook.com/intern/diff/D26261042/)

[ghstack-poisoned]
